File: frontend/src/interfaces/im8.ts
import { DatabaseProps } from ".";
import type { MitigationMeasure } from "./register";

//////////////////////////////////////////////////
// IM8 Table

//////////////////////////////////////////////////
export interface KbIM8 extends DatabaseProps {
    im8: IM8Policy[];
}

export interface IM8Policy {
    id: string;
    //
    annex_ref: string;
    category: string;
    domain: string;
    footnote_ref: string;
    guideline_ref: string;
    image_name: string;
    image_url: string;
    policy_ref: string;
    std_ref: string;
    subdomain: string;
    tags: string;
    text: string;
    // Mitigation measures associated with this policy (backend-provided).
    mitigations?: MitigationMeasure[];
}

File: backend/modules/shared_libs/src/shared_libs/models/base_models/im8.py
from typing import Self

from pydantic import BaseModel, Field, model_validator

from shared_libs.models.base_models.shared.shared.database import PatchBaseModel
from shared_libs.models.model_validators import KbIM8BaseValidator

__all__ = [
    "IM8PolicyModel",
    "KbIM8BaseModel",
]


class IM8PolicyModel(BaseModel):
    id: str
    annex_ref: str | None = Field(default="")
    category: str | None = Field(default="")
    domain: str | None = Field(default="")
    footnote_ref: str | None = Field(default="")
    guideline_ref: str | None = Field(default="")
    image_name: str | None = Field(default="")
    image_url: str | None = Field(default="")
    policy_ref: str | None = Field(default="")
    in_scope: bool | None = Field(default=False)
    is_auto_compliant: bool | None = Field(default=False)
    std_ref: str | None = Field(default="")
    subdomain: str | None = Field(default="")
    tags: str | None = Field(default="")
    text: str | None = Field(default="")


class KbIM8BaseModel(
    PatchBaseModel,
    KbIM8BaseValidator,
):
    im8: list["IM8PolicyModel"] | None = Field(default=[])

    @model_validator(mode="wrap")
    @classmethod
    def validate_model(
        cls,
        data: dict,
        handler,
    ) -> Self:
        model = cls.get_validated_model(
            data=data,
            handler=handler,
        )
        return model
